core/src/main/java/org/elasticsearch/index/seqno/LocalCheckpointTracker.java

```diff
@@ -183,63 +175,70 @@ synchronized void waitForOpsToComplete(final long seqNo) throws InterruptedExcep
+            if (current == null) {
+                // the bit set corresponding to the checkpoint has already been removed, set ourselves up for the next bit set
+                assert checkpoint % bitArraysSize == bitArraysSize - 1;
+                current = processedSeqNo.get(++bitArrayKey);
+            }
             do {
                 checkpoint++;
-                // the checkpoint always falls in the first bit set or just before. If it falls
-                // on the last bit of the current bit set, we can clean it.
-                if (checkpoint == firstProcessedSeqNo + bitArraysSize - 1) {
-                    processedSeqNo.removeFirst();
-                    firstProcessedSeqNo += bitArraysSize;
-                    assert checkpoint - firstProcessedSeqNo < bitArraysSize;
-                    current = processedSeqNo.peekFirst();
+                /*
+                 * The checkpoint always falls in the current bit set or we have already cleaned it; if it falls on the last bit of the
+                 * current bit set, we can clean it.
+                 */
+                if (checkpoint == lastSeqNoInBitArray(bitArrayKey)) {
+                    assert current != null;
+                    final FixedBitSet removed = processedSeqNo.remove(bitArrayKey);
+                    assert removed == current;
+                    current = processedSeqNo.get(++bitArrayKey);
                 }
-            } while (current != null && current.get(seqNoToBitSetOffset(checkpoint + 1)));
+            } while (current != null && current.get(seqNoToBitArrayOffset(checkpoint + 1)));
         } finally {
             // notifies waiters in waitForOpsToComplete
             this.notifyAll();
         }
     }
 
+    private long lastSeqNoInBitArray(final long bitArrayKey) {
+        return (1 + bitArrayKey) * bitArraysSize - 1;
+    }
+
     /**
      * Return the bit array for the provided sequence number, possibly allocating a new array if needed.
      *
      * @param seqNo the sequence number to obtain the bit array for
      * @return the bit array corresponding to the provided sequence number
      */
-    private FixedBitSet getBitSetForSeqNo(final long seqNo) {
+    private long getBitArrayKey(final long seqNo) {
         assert Thread.holdsLock(this);
-        assert seqNo >= firstProcessedSeqNo : "seqNo: " + seqNo + " firstProcessedSeqNo: " + firstProcessedSeqNo;
-        final long bitSetOffset = (seqNo - firstProcessedSeqNo) / bitArraysSize;
-        if (bitSetOffset > Integer.MAX_VALUE) {
-            throw new IndexOutOfBoundsException(
-                "sequence number too high; got [" + seqNo + "], firstProcessedSeqNo [" + firstProcessedSeqNo + "]");
-        }
-        while (bitSetOffset >= processedSeqNo.size()) {
-            processedSeqNo.add(new FixedBitSet(bitArraysSize));
+        return seqNo / bitArraysSize;
+    }
+
+    private FixedBitSet getBitArrayForSeqNo(final long seqNo) {
```

Comment:
maybe use `indexOf(...)` and then `indexInsert(...)` and `indexGet(...)` respectively to avoid determining what the slot is for a key several times?

```java
final int slot = processedSeqNo.indexOf(bitArrayKey);
if (processedSeqNo.indexExists(slot) == false) {
   processedSeqNo.indexInsert(slot, bitArrayKey, new FixedBitSet(bitArraysSize));
}
return processedSeqNo.indexGet(slot);
```

Reply:
Thanks @martijnvg, I pushed cabda878aad226d0fae15331ff0ea5e5c792cf98.

Reply:
And 68c4eab0f50cf615a705ab8ccaa6efd31d121c1e. 😐

Reply:
Actually the `indexGet` at the end is not correct because the returned index is negative if the slot is not actually used. Yet, `~slot` also does not work because the table could have been resized after the insert. I pushed 97bb3cadbf97da876ee305243da9bc26e81e9e98.

Reply:
ah, I didn't realize that. Maybe we should do this to make optimal use of the `slot`?

```
final int slot = processedSeqNo.indexOf(bitArrayKey);
if (processedSeqNo.indexExists(slot) == false) {
   FixedBitSet bitSet = new FixedBitSet(bitArraysSize));
   processedSeqNo.indexInsert(slot, bitArrayKey, bitSet);
   return bitSet;
} else {
   return processedSeqNo.indexGet(slot);
}
```

Reply:
When you see a chance to avoid `== false`, you seize it. 😉

I pushed f4f0dae6043726600d5cd7ad65253879a7df67fc.

Reply:
:)